core/src/deno/script.rs

```diff
@@ -0,0 +1,147 @@
+use anyhow::Result;
+use deno_core::{op, Extension, JsRuntime, OpState, ZeroCopyBuf};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::borrow::Cow;
+use std::rc::Rc;
+use std::{thread, time::Duration};
+
+pub struct Script {
+    runtime: JsRuntime,
+    last_rid: u32,
+    timeout: Option<Duration>,
+}
+
+impl Script {
+    const DEFAULT_FILENAME: &'static str = "code_block.js";
+
+    pub fn from_string(js_code: &str) -> Result<Self> {
+        // console.log() is not available by default -- add the most basic version with single
+        // argument (and no warn/info/... variants)
+        let all_code =
+            "const console = { log: function(expr) { Deno.core.print(expr + '\\n', false); } };"
+                .to_string()
+                + js_code;
+
+        Self::create_script(all_code)
+    }
+
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        assert!(self.timeout.is_none());
+        assert!(timeout > Duration::ZERO);
+
+        self.timeout = Some(timeout);
+        self
+    }
+
+    fn create_script(js_code: String) -> Result<Self> {
+        let ext = Extension::builder("script")
+            .ops(vec![(op_return::decl())])
+            .build();
+
+        let options = deno_core::RuntimeOptions {
+            module_loader: Some(Rc::new(deno_core::FsModuleLoader)),
+            extensions: vec![ext],
+            ..Default::default()
+        };
+
+        let mut runtime = JsRuntime::new(options);
+
+        // We cannot provide a dynamic filename because execute_script() requires a &'static str
+        runtime.execute_script(Self::DEFAULT_FILENAME, js_code.into())?;
+
+        Ok(Script {
+            runtime,
+            last_rid: 0,
+            timeout: None,
+        })
+    }
+
+    pub fn call<A, R>(&mut self, fn_name: &str, arg: A) -> Result<R>
+    where
+        A: Serialize,
+        R: DeserializeOwned,
+    {
+        let json_arg = serde_json::to_value(arg)?.to_string();
+        let json_result = self.call_impl(fn_name, json_arg)?;
+        let result: R = serde_json::from_value(json_result)?;
+
+        Ok(result)
+    }
+
+    fn call_impl(&mut self, fn_name: &str, json_args: String) -> Result<serde_json::Value> {
+        // Note: ops() is required to initialize internal state
+        // Wrap everything in scoped block
+
+        // 'undefined' will cause JSON serialization error, so it needs to be treated as null
+        let js_code = format!(
+            "(async () => {{
+				let __rust_result = {fn_name}.constructor.name === 'AsyncFunction'
+					? await {fn_name}({json_args})
+					: {fn_name}({json_args});
+
+				if (typeof __rust_result === 'undefined')
+					__rust_result = null;
+
+				Deno.core.ops.op_return(__rust_result);
+			}})()"
+        );
+
+        // println!("CALLING SCRIPT:\n{}", js_code);
+
+        if let Some(timeout) = self.timeout {
+            let handle = self.runtime.v8_isolate().thread_safe_handle();
```

Comment:
More of a question, but do you know how much overhead it adds to run a full deno for each script ?

Reply:
Yeah unclear. Not a full deno it's a JsRuntime so a simple wrapper around v8